### PR TITLE
JAMES-2366 Group prefix & refactorings

### PR DESCRIPTION
**Pull request by mbaechler**
This comes as a replacement for https://github.com/linagora/james-project/pull/1371
Rebased on current master of 2 hours of cherry-pick and conflict solving !
Please review it all.

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[5864aff2ac267091e51a912b1335f8dcb586fbf8] Failed tests: 
[5864aff2ac267091e51a912b1335f8dcb586fbf8]   AbstractRecipientRewriteTableTest.getAllMappingsShouldListAllEntries:126 [One mappingline] 
[5864aff2ac267091e51a912b1335f8dcb586fbf8] Expecting:
[5864aff2ac267091e51a912b1335f8dcb586fbf8]   <{org.apache.james.rrt.lib.MappingSource@19ea1f5e=MappingsImpl{mappings=[Mapping{type=Regex mapping=(.*)@localhost}, Mapping{type=Regex mapping=(.+)@test}, Mapping{type=Address mapping=test@localhost}]}}>
[5864aff2ac267091e51a912b1335f8dcb586fbf8] to contain only:
[5864aff2ac267091e51a912b1335f8dcb586fbf8]   <[(org.apache.james.rrt.lib.MappingSource@19ea1f5e,MappingsImpl{mappings=[Mapping{type=Regex mapping=(.*)@localhost}, Mapping{type=Regex mapping=(.+)@test}]}),
[5864aff2ac267091e51a912b1335f8dcb586fbf8]     (org.apache.james.rrt.lib.MappingSource@19ea1f5e,MappingsImpl{mappings=[Mapping{type=Address mapping=test@localhost}]})]>
[5864aff2ac267091e51a912b1335f8dcb586fbf8] elements not found:
[5864aff2ac267091e51a912b1335f8dcb586fbf8]   <[MapEntry[key=org.apache.james.rrt.lib.MappingSource@19ea1f5e, value=MappingsImpl{mappings=[Mapping{type=Address mapping=test@localhost}]}]]>
[5864aff2ac267091e51a912b1335f8dcb586fbf8] and elements not expected:
[5864aff2ac267091e51a912b1335f8dcb586fbf8]   <[MapEntry[key=org.apache.james.rrt.lib.MappingSource@19ea1f5e, value=MappingsImpl{mappings=[Mapping{type=Regex mapping=(.*)@localhost}, Mapping{type=Regex mapping=(.+)@test}, Mapping{type=Address mapping=test@localhost}]}]]>
```

**Comment by rouazana** [comment]
sorry, it seems I missed to submit my review yesterday: https://github.com/linagora/james-project/pull/1371#pullrequestreview-113611790
I will also review this, but can you have a look at my comments please?

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java [hunk 1]
```diff
@@ -77,6 +78,40 @@ static IdentityMappingPolicy selectIdentityPolicy(Type type) {
         throw new IllegalStateException("unhandle enum type");
     }
 
+    enum MailAddressConversionPolicy {
```
Comment (aduprat) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: Maybe a functional interface is more accurate to this goal.

File: server/data/data-api/src/test/java/org/apache/james/rrt/lib/MappingTest.java [hunk 2]
```diff
@@ -139,4 +143,167 @@ public void withoutPrefixShouldThrowWhenNoPrefixOnForwardType() {
         assertThatThrownBy(() -> Type.Forward.withoutPrefix("mapping"))
             .isInstanceOf(IllegalArgumentException.class);
     }
+
+
```
Comment (aduprat) [inline server/data/data-api/src/test/java/org/apache/james/rrt/lib/MappingTest.java]: extra line

File: core/src/main/java/org/apache/james/core/User.java [hunk 3]
```diff
@@ -84,11 +92,31 @@ public boolean hasDomainPart() {
...
+    public User withDefaultDomain(Domain defaultDomain) {
+        if (hasDomainPart()) {
+            return this;
+        }
+        return new User(localPart, Optional.of(defaultDomain));
+    }
+
+    public User withDefaultDomainFromUser(User other) {
+        Preconditions.checkArgument(other.hasDomainPart());
+        if (hasDomainPart()) {
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply core/src/main/java/org/apache/james/core/User.java]: yes

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 4]
```diff
@@ -94,124 +95,112 @@ public void setMappingLimit(int mappingLimit) throws IllegalArgumentException {
...
+        Mappings childMappings = getMappings(targetUser, remainingLoops - 1);
+
+        if (childMappings.isEmpty()) {
+            return nonRecursiveResult;
+        } else {
+            return childMappings.asStream();
+        }
+    }
+
+    private Mapping toMapping(User rewrittenUser, Type type) {
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: I won't go further in this refactoring, sorry

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java [hunk 5]
```diff
@@ -201,10 +205,10 @@ private void ensureNotShadowingAnotherAddress(MailAddress groupAddress) throws U
     public HaltException removeFromGroup(Request request, Response response) throws JsonExtractException, AddressException, RecipientRewriteTableException {
         MailAddress groupAddress = parseMailAddress(request.params(GROUP_ADDRESS));
         MailAddress userAddress = parseMailAddress(request.params(USER_ADDRESS));
-        recipientRewriteTable.removeAddressMapping(
-            groupAddress.getLocalPart(),
-            groupAddress.getDomain(),
-            userAddress.asString());
+        MappingSource source = MappingSource
+            .fromUser(
+                User.fromLocalPartWithDomain(groupAddress.getLocalPart(), groupAddress.getDomain()));
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: we could, could we do that in another PR ?

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java (hunk 1 above)
Comment (mbaechler) [reply server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: Using an enum allows to : 
* limit the number of implementation
* name the functions
It better suites this case

File: server/data/data-api/src/test/java/org/apache/james/rrt/lib/MappingTest.java (hunk 2 above)
Comment (mbaechler) [reply server/data/data-api/src/test/java/org/apache/james/rrt/lib/MappingTest.java]: ok

**Comment by mbaechler** [comment]
```
[a2edff43ba3a2a8f772f24577ed9cb513237fdb4] Tests in error: 
[a2edff43ba3a2a8f772f24577ed9cb513237fdb4]   DataCommandsIntegrationTest.listMappingsShouldWork:158 » Unmarshal error unmar...
[a2edff43ba3a2a8f772f24577ed9cb513237fdb4]   DataCommandsIntegrationTest.listUsersDomainMappingShouldWork:170 » Unmarshal e...
```

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/MappingSource.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    public Optional<Domain> asDomain() {
+        return domain;
+    }
+
+    public Optional<MailAddress> asMailAddress() {
+        return user.flatMap(user -> {
+            try {
+                return Optional.of(user.asMailAddress());
+            } catch (AddressException e) {
+                LOGGER.warn("ignoring failing User to MailAddress convertion");
```
Comment (rouazana) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/MappingSource.java]: add the failing user in the log message?

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/MappingSource.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    public String getFixedUser() {
+        return user.map(User::getLocalPart)
+            .orElse(WILDCARD);
+    }
+
+    public String getFixedDomain() {
+        return OptionalUtils.or(
+            user.flatMap(User::getDomainPart).map(Domain::asString),
+            domain.map(Domain::asString))
+            .orElse(WILDCARD);
```
Comment (rouazana) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/MappingSource.java]: indent?

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/RecipientRewriteTableUtil.java [hunk 8]
```diff
@@ -46,32 +46,11 @@ private RecipientRewriteTableUtil() {
...
-        if (target != null) {
-            return target;
-        }
-
-        return null;
+    public static String getTargetString(String user, Domain domain, Map<MappingSource, String> mappings) {
+        return OptionalUtils.or(
+            Optional.ofNullable(mappings.get(MappingSource.fromUser(user, domain))),
+            Optional.ofNullable(mappings.get(MappingSource.fromDomain(domain))))
+            .orElse(null);
```
Comment (rouazana) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/RecipientRewriteTableUtil.java]: indent?

File: core/src/main/java/org/apache/james/core/User.java [hunk 9]
```diff
@@ -99,11 +99,7 @@ public User withDefaultDomain(Domain defaultDomain) {
     }
 
     public User withDefaultDomainFromUser(User other) {
-        Preconditions.checkArgument(other.hasDomainPart());
-        if (hasDomainPart()) {
-            return this;
-        }
-        return new User(localPart, Optional.of(other.domainPart.get()));
+        return withDefaultDomain(other.domainPart.get());
```
Comment (rouazana) [inline core/src/main/java/org/apache/james/core/User.java]: this is a public API for User class. You cannot assume other will have a domain, and we can considerate as legit for other to not have a domain. I think you should you get domain from other (empty or non empty) without wondering if it has a domain

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: Is it long to implement?

[the remark this answers is not in the document]
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/GroupsRoutes.java]: that's not the point. The PR is already too big, we should take the habit to handle improvements in others PRs too merge faster

**Comment by mbaechler** [comment]
merged

**Pull request closed without merging**